[RFC] vim-patch:8.0.1807,8.1.{453,454}

janlazo opened the pull request:
**vim-patch:8.0.1807: function to set terminal name is too long**

Problem:    Function to set terminal name is too long.
Solution:   Refactor the function.  Fix typo in test.
https://github.com/vim/vim/commit/69e056915c4145b7b64c60963797692a5b293561

**vim-patch:8.1.0453: MS-Windows: executable() is not reliable**

Problem:    MS-Windows: executable() is not reliable.
Solution:   Use $PATHEXT properly. (Yasuhiro Matsumoto, closes vim/vim#3412)
vim/vim@8295666

**vim-patch:8.1.0454: resolve() was not tested with a symlink cycle**

Problem:    resolve() was not tested with a symlink cycle.
Solution:   Add a test. (Dominique Pelle, closes vim/vim#3513)
vim/vim@2610990

janlazo committed:
vim-patch:8.0.1807: function to set terminal name is too long

Problem:    Function to set terminal name is too long.
Solution:   Refactor the function.  Fix typo in test.
https://github.com/vim/vim/commit/69e056915c4145b7b64c60963797692a5b293561

janlazo committed:
vim-patch:8.1.0454: resolve() was not tested with a symlink cycle

Problem:    resolve() was not tested with a symlink cycle.
Solution:   Add a test. (Dominique Pelle, closes vim/vim#3513)
https://github.com/vim/vim/commit/261099070940fb64d4633a766ae3510fffe4171d

janlazo committed:
vim-patch:8.1.0453: MS-Windows: executable() is not reliable

Problem:    MS-Windows: executable() is not reliable.
Solution:   Use $PATHEXT properly. (Yasuhiro Matsumoto, closes vim/vim#3412)
https://github.com/vim/vim/commit/8295666dc2c65e42135b91d5c61e2a140d002333

janlazo commented on src/nvim/testdir/test_functions.vim:
```diff
@@ -818,6 +864,19 @@ func Test_filewritable()
+    call assert_equal(1, executable('notepad.exe'))
+    call assert_equal(0, executable('notepad.exe.exe'))
+    call assert_equal(0, executable('shell32.dll'))
+    call assert_equal(0, executable('win.ini'))
```
Why are these not executable? `dll,ini` don't open notepad?

```
Found errors in Test_Executable():
function RunTheTest[35]..Test_Executable line 5: Expected 0 but got 1
function RunTheTest[35]..Test_Executable line 6: Expected 0 but got 1
```

janlazo replied:
Forgot that `executable` checks files in `PATH`. shell32.dll is in `PATH` but `dll` is not executable (won't open notepad as fallback).  `win.ini` is in `PATH` (C:\Windows\win.ini on Windows 8.1) and opens notepad so it is executable. notepad.exe.exe is not in `PATH` (unless it was created elsewhere) so it is not executable here.

janlazo replied:
`win.ini` is loaded to notepad because of the default file assocation between `inifile` and `notepad` on Windows.
In cmd.exe, user can associate file extensions with a filetype and a default command via [assoc](https://ss64.com/nt/assoc.html) and [ftype](https://ss64.com/nt/ftype.html) commands.

justinmk commented:
Do you want to merge this, or are you planning  to check/fix the PATHEXT logic (if needed)?

janlazo commented:
Not sure what to fix. If there's a fix, then it's a file permission issue like in Unix. Using File Explorer to check file properties dialog, there is Security tab that shows the file permissions. One of which is `Read & execute` permission, separate from `Read` permission. Fix is to check that execute permission (is that handled by libuv?).

`PATHEXT` allows the user to "execute" files while omitting the file extensions found in `assoc` (or the equivalent registry check). `assoc` associates file extension with a filetype which `ftype` uses to run commands based on the filetype. `assoc` and `ftype` are internal cmd.exe commands.
What's strange is that I can open unassociated file extensions with notepad and dll files are not "executable" because `ftype` does not handle dll files.
User can fix this by running dll files with [run32dll](https://ss64.com/nt/rundll32.html) (as I do in my [vimrc](https://github.com/janlazo/dotvim8/commit/8ce2325798389a80d4689060d38dc9f024c98803#diff-a7d4f465a716e696f0a036380c50d213R40))
or by appending `;.dll` to `PATHEXT` to open dll files with notepad.
As long as the user can "read and execute" the file, then the file is "executable".
Perhaps `PATHEXT` is for the shell only and powershell supports it for backward compatibility with cmd.exe. `.ps1` is not in `PATHEXT` by default but it is executable in powershell.

justinmk commented:
We deal with PATHEXT in `os_can_exe`, but `is_executable` pretty much always returns true, so `is_executable_ext` is pointless.


> Perhaps PATHEXT is for the shell only and powershell supports it for backward compatibility with cmd.exe. .ps1 is not in PATHEXT by default but it is executable in powershell

Ah, good point.  

Anyways,  it looks like `call assert_equal(0, executable('win.ini'))` is failing in the AppVeyor log. Shall we just change it to `1` for now?

`executable()` on Windows has been useless for decades, and I wouldn't be surprised if 8.1.0454 gets reverted. It seems like it was rushed.

janlazo commented:
> Anyways, it looks like call assert_equal(0, executable('win.ini')) is failing in the AppVeyor log. Shall we just change it to 1 for now?

Sure.

janlazo committed:
oldtests: win: fix executable() assertions

Windows has "Read and execute" permission via ACL
but nvim and libuv do not support ACL.
Windows does not support the executable bit in chmod-style permissions
but it is safe to assume that if the file exists and is readable,
then it is most likely executable.
This means that win.ini and shell32.dll are "executable"
because they exist, are readable, and are in PATH.

PATHEXT does not affect the executable permission of a file;
it exists to run files on the shell while omitting the file extension.
Assume that PATHEXT is intended for cmd.exe only
because powershell can execute powershell files (ie. *.ps1)
without changing PATHEXT or related cmd.exe environment variable.

In the future, nvim should check the outputs of 'assoc' and 'ftype',
cmd.exe internal commands, or check the registry.
Powershell can be used for ACL if C++/C# API is too difficult to use.
